### DISCUSSION TITLE
Improving namespace structure

### DISCUSSION THREAD

**Discussion by cflaviu** [managarm/managarm#307] [Ideas and RFCs]
The namespace structure could be improved because many related namespaces are not grouped. The namespaces could be structured following the folder structure of the project.

**Comment by 64** [comment]
Do you have an example in mind?

**Reply by cflaviu** [reply]
eg. src/managarm/drivers/kbd/ -> namespace managarm::driver::kbd

**Reply by ArsenArsen** [reply]
I'm not in favor of excessive verbosity in these, fwiw, e.g. `managarm::` seems useless outside of libraries, and codebases that are not linked together shouldn't really influence eachothers namespacing either (by "inducing" namespaces where they would otherwise not be)

**Reply by cflaviu** [reply]
@ArsenArsen The big C++ projects I know are using namespaces consistently following project modules/functionalities according to software engineering guides. Independent (reusable) libraries should have their own root namespace. Modern C++ allows nested namespace definitions and therefore verbosity is low. At lease for me, or for newcomers, it is easier to follow the code using Class View. It is hard doing this in the current state. In Class View I would expect to identify easily the 3rd party libraries and the modules of the project.

**Reply by avdgrinten** [reply]
I think `driver::kbd` is too excessive but namespacing individual components probably still makes sense. Something like `managarm::ps2_hid` or `managarm::block_ata` sounds reasonable to me.

The `managarm` namespace is not painful if everything is in `managarm` since you can then drop the qualification.

**Reply by cflaviu** [reply]
@avdgrinten Would you feel comfortable on a new project navigating through the classes as they are displayed in class tree as in the linked picture (below) + 2 more scroll pages? A too flat hierarchy could also be problematic.

**Reply by avdgrinten** [reply]
There are not *that* many components in Managarm (yet), so this would not result in multiple pages of namespaces. Besides, the is always the option to use the file system hierarchy to browse the project.

**Reply by cflaviu** [reply]
@avdgrinten I can put all namespace listed in Class View in a picture. It's big.

**Reply by cflaviu** [reply]
It's easier making such changes at an earlier stage than changing later having much more code.

**Reply by cflaviu** [reply]
> I think `driver::kbd` is too excessive but namespacing individual components probably still makes sense. Something like `managarm::ps2_hid` or `managarm::block_ata` sounds reasonable to me.
> 
> The `managarm` namespace is not painful if everything is in `managarm` since you can then drop the qualification.

Well you prefer a more flat namespace hierarchy but contrarily you prefer more nested file system structure of the project. Having same approach structuring namespace and folder structure in the same way, such consistency, will help understanding the project and finding items easier.

**Comment by cflaviu** [comment]
In Class View these namespaces appear https://pasteboard.co/K3miF1s.jpg on the first scroll page. There are more scroll pages.

**Comment by avdgrinten** [comment]
I think this would be a good addition, but it would be good to first propose a list of namespaces before we apply the changes / before somebody works on patches.

**Reply by cflaviu** [reply]
We could create a wiki page where we can collaborate on namespace structure.

**Comment by cflaviu** [comment]
```
arch
async
bragi
cralgo
frigg
managarm
	core
		drm
		virt_io
	driver
		block
		clock_tracker
		gfx
		kbd
		kernlet
		nic
		tty
		uart
		usb
	eir
	helix
	mbus
	posix
	protocol
		fs
		hw
		kernlet
		mbus
		os_trace
		posix
		srv_ctl
		usb
	server
		netserver
	thor
	utils
		lsm_bus
		run_srv
smarter

```
An example.
